sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/KeyMetaDataStore.java

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.android.communication.chat.implementation.notifications.fcm;
+
+import com.azure.android.core.logging.ClientLogger;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class represents a storage facility for storing alias and its creation time of secret keys. The
+ * data is persisted in a file as json format. The key-store itself could load data from file for
+ * faster operation.
+ *
+ * The file persists all the data in Json format. Latest record is added to last of the list. For example:
+ * {
+ *     "key_1": {
+ *         iv: 61 76 69 66 70 6f 69 77 69 63,
+ *         encryptionText: 61 76 69 66 70 6f 69 77 69 64,
+ *         creationTime: 1234
+ *     },
+ *     "key_2": {
+ *         iv: 61 76 69 66 70 6f 69 77 69 11,
+ *         encryptionText: 61 76 69 66 70 6f 69 77 69 33,
+ *         creationTime: 1234
+ *     }
+ *
+ * }
+ */
+public class KeyMetaDataStore {
+    private Map<String, KeyMetaDataEntry> map;
+
+    private ClientLogger clientLogger = new ClientLogger(KeyMetaDataStore.class);
+
+    public KeyMetaDataStore() {
+        map = new HashMap<>();
+    }
+
+    public int getSize() {
+        return map.size();
+    }
+
+    //Loading data with given path to memory
+    public void load(InputStream inputStream) {
+        //Nothing to read from
+        if (inputStream == null) {
+            return;
+        }
+
+        // convert JSON array to map
+        try {
+            map = new ObjectMapper().readValue(inputStream, new TypeReference<Map<String, KeyMetaDataEntry>>() { });
+        } catch (IOException e) {
+            clientLogger.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    //Write in-memory map into file with Json format
+    private void writeJsonToFile(String path) {
+        File outputFile = new File(path);
+        try {
+            boolean newFile = outputFile.createNewFile();
+            if (newFile) {
+                clientLogger.verbose("new file created for storing push notification credentials");
+            }
+        } catch (IOException e) {
+            clientLogger.logExceptionAsError(new RuntimeException("Failed to create key store file", e));
```

Comment:
If we reach this catch block i assume one reason could be because the file already exists, right? What about other scenarios? Is there a case where the file does not exist and could not be created? (e.g. having an invalid path).

I ask so that we don't carry on with the logic below if there is no file available. I guess a catch block further down could handle that scenario...

Reply:
file.createNewFile() operation returns false instead of throwing an exception if file already exist. So we just need handle IOException for exception during creating of a file. But I will change to throw an exception in the catch.